istioctl/cmd/injector-list.go

```diff
@@ -0,0 +1,328 @@
+
+package cmd
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"sort"
+	"strconv"
+	"strings"
+	"text/tabwriter"
+
+	"github.com/spf13/cobra"
+	admit_v1 "k8s.io/api/admissionregistration/v1"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	api_pkg_labels "k8s.io/apimachinery/pkg/labels"
+
+	"istio.io/api/annotation"
+	"istio.io/api/label"
+	"istio.io/istio/galley/pkg/config/analysis/analyzers/injection"
+	analyzer_util "istio.io/istio/galley/pkg/config/analysis/analyzers/util"
+	"istio.io/istio/istioctl/pkg/clioptions"
+	"istio.io/istio/pkg/config/resource"
+	"istio.io/istio/pkg/kube"
+)
+
+type revisionCount struct {
+	// pods in a revision
+	pods int
+	// pods that are disabled from injection
+	disabled int
+	// pods that are enabled for injection, but whose revision doesn't match their namespace's revision
+	needsRestart int
+}
+
+func injectorCommand() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:     "injector",
+		Short:   "List sidecar injector and sidecar versions",
+		Long:    `List sidecar injector and sidecar versions`,
+		Example: `  istioctl experimental injector list`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			cmd.HelpFunc()(cmd, args)
+			if len(args) != 0 {
+				return fmt.Errorf("unknown subcommand %q", args[0])
+			}
+
+			return nil
+		},
+	}
+
+	cmd.AddCommand(injectorListCommand())
+	return cmd
+}
+
+func injectorListCommand() *cobra.Command {
+	var opts clioptions.ControlPlaneOptions
+	cmd := &cobra.Command{
+		Use:     "list",
+		Short:   "List sidecar injector and sidecar versions",
+		Long:    `List sidecar injector and sidecar versions`,
+		Example: `  istioctl experimental injector list`,
+		RunE: func(cmd *cobra.Command, args []string) error {
+			client, err := kubeClientWithRevision(kubeconfig, configContext, opts.Revision)
+			if err != nil {
+				return fmt.Errorf("failed to create k8s client: %v", err)
+			}
+			nslist, err := getNamespaces(client)
+			if err != nil {
+				return err
+			}
+			hooks, err := getWebhooks(client)
+			if err != nil {
+				return err
+			}
+			pods, err := getPods(client)
+			if err != nil {
+				return err
+			}
+			err = printNS(cmd.OutOrStdout(), nslist, hooks, pods)
+			if err != nil {
+				return err
+			}
+			cmd.Println()
+			injectedImages, err := getInjectedImages(client)
+			if err != nil {
+				return err
+			}
+			return printHooks(cmd.OutOrStdout(), nslist, hooks, injectedImages)
+		},
+	}
+
+	return cmd
+}
+
+func getNamespaces(client kube.ExtendedClient) ([]v1.Namespace, error) {
+	nslist, err := client.CoreV1().Namespaces().List(context.TODO(), metav1.ListOptions{})
+	if err != nil {
+		return []v1.Namespace{}, err
+	}
+	return nslist.Items, nil
+}
+
+func printNS(writer io.Writer, namespaces []v1.Namespace, hooks []admit_v1.MutatingWebhookConfiguration,
+	allPods map[resource.Namespace][]v1.Pod) error {
+
+	sort.Slice(namespaces, func(i, j int) bool {
+		return namespaces[i].Name < namespaces[j].Name
+	})
```

Comment:
nit: I think this makes more sense outside of printNS. At some point we'll probably add different output formats and we probably want all of them sorted.